cli/cmd/install.go

```diff
@@ -190,40 +161,50 @@ Otherwise, you can use the --ignore-cluster flag to overwrite the existing globa
 // These options may be overridden on the CLI at install-time and will be
 // persisted in Linkerd's control plane configuration to be used at
 // injection-time.
-func newInstallOptionsWithDefaults() *installOptions {
+func newInstallOptionsWithDefaults() (*installOptions, error) {
+	chartDir := fmt.Sprintf("%s%s", helmDefaultChartDir, string(filepath.Separator))
+	defaults, err := charts.ReadDefaults(chartDir, false)
+	if err != nil {
+		return nil, err
+	}
+
 	return &installOptions{
 		controlPlaneVersion:         version.Version,
-		controllerReplicas:          defaultControllerReplicas,
-		controllerLogLevel:          "info",
+		controllerReplicas:          defaults.ControllerReplicas,
+		controllerLogLevel:          defaults.ControllerLogLevel,
 		highAvailability:            false,
-		controllerUID:               2103,
-		disableH2Upgrade:            false,
+		controllerUID:               defaults.ControllerUID,
+		disableH2Upgrade:            !defaults.EnableH2Upgrade,
 		noInitContainer:             false,
-		omitWebhookSideEffects:      false,
+		omitWebhookSideEffects:      defaults.OmitWebhookSideEffects,
 		restrictDashboardPrivileges: false,
 		proxyConfigOptions: &proxyConfigOptions{
 			proxyVersion:           version.Version,
 			ignoreCluster:          false,
-			proxyImage:             defaultDockerRegistry + "/proxy",
-			initImage:              defaultDockerRegistry + "/proxy-init",
+			proxyImage:             defaults.ProxyImageName,
+			initImage:              defaults.ProxyInitImageName,
 			initImageVersion:       version.ProxyInitVersion,
 			dockerRegistry:         defaultDockerRegistry,
```

Comment:
curious: why do fields like `initImageVersion` and `dockerRegistry` not come from `charts.ReadDefaults()`? what's the distinction between a default hard-coded here vs. `charts.ReadDefaults()`

Reply:
Good questions. `initImageVersion` and `ProxyVersion` remain as-is because their values need to be updated by the `version` package across builds. These values aren't returned by `charts.ReadDefaults()`, because the static versions in the `values.yaml` and `values-ha.yaml` files are only relevant to `helm install`.

`dockerRegistry` remains as-is because it isn't configurable by Helm user. (Hence, it isn't specified in the `values.yaml` file.) The reason being the way we structured the `installValues` in our original Helm template. E.g. on [`master`](https://github.com/linkerd/linkerd2/blob/b4b2a44299093cf5ad24f046ce3d4bf423b1cf68/chart/templates/controller.yaml#L75), only `{{ControllerImage}}` is configurable by the `install` code, with the docker registry baked into that variable.

I'm open to suggestions on how to make this better and clearer.

Reply:
I've created #3237 to revisit this, post 2.5. Ideally, the `pkg/charts` library should provide a clean interface for the CLI and webhook to work with these values types.